test/test_charm.py

```diff
@@ -15,9 +17,14 @@ class TestCharm(unittest.TestCase):
     def setUp(self):
         self.tmpdir = Path(tempfile.mkdtemp())
 
+        # Avoid introducing side-effects for other tests by using a separate type.
```

Comment:
This should be more clear about which side effects it's talking about, otherwise we'll have to go after boring details to figure it out again in a few months:
```
# Relations events are defined dynamically and modify the class attributes.
# We use a subclass temporarily to prevent these side effects from leaking. 
```